### PR TITLE
feat: Writing assistant

### PR DESCRIPTION
**Pull request by weilirs**
Reflect-like AI writing assistant

- 3 system prompts(simplify, copy editor, takeaways)
- users can specify their own prompts
- users can copy, insert the LLM response, or replace the highlighted text with LLM response

Now the prompts are not perfect, I think they need more prompt engineering.

Demo:

https://github.com/user-attachments/assets/8603bfd7-447c-4ca4-ac29-b6c8891909c2

### CODE REVIEW

File: src/components/Writing-Assistant/WritingAssistantFloatingMenu.tsx [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+      currentChatHistory.displayableChatHistory[
+        currentChatHistory.displayableChatHistory.length - 1
+      ];
+
+    const copiedText = llmResponse.visibleContent
+      ? llmResponse.visibleContent
+      : formatOpenAIMessageContentIntoString(llmResponse.content);
+
+    navigator.clipboard
+      .writeText(copiedText)
```
Comment (samlhuillier) [inline src/components/Writing-Assistant/WritingAssistantFloatingMenu.tsx:76]: Please fix these build issues

File: src/components/Writing-Assistant/WritingAssistantFloatingMenu.tsx [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+""" ${selectedText} """
+Write a markdown list (using dashes) of key takeaways from my notes. Write at least 3 items, but write more if the text requires it. Be very detailed and don't leave any information out. Do not wrap responses in quotes.`;
+        break;
+      case "custom":
+        prompt =
+          `prompt(in triple #): ` +
+            `### ` +
+            customPromptInput +
+            ` ###` +
+            ` ,apply the prompt to the text in triple quotes """ ${selectedText} """` ||
```
Comment (samlhuillier) [inline src/components/Writing-Assistant/WritingAssistantFloatingMenu.tsx:179]: This prompt imo could be improved to be like "The user has given the following instructions for processing the text selected" or something like htat

**Comment by samlhuillier** [comment]
This is absolutely excellent!
Couple of points after running it:
- We need to handle the case where the generated text flows off the screen. Atm it just goes off the screen unless the window is resized. I think a good solution is to calculate the max size of the component based on the screen height and current position of the component and then once the component reaches this max size just make it have scrollbars in the text it has generated. Thouh overall I think this could be added in a future PR.
- Another follow on from this will be to think about just like the slash AI command (where no context needs to be selected) which can be used to just write or take in the context of the whole note. Again work for a future PR!

File: src/components/Writing-Assistant/WritingAssistantFloatingMenu.tsx (hunk 2 above)
Comment (weilirs) [reply src/components/Writing-Assistant/WritingAssistantFloatingMenu.tsx]: modified

File: src/components/Writing-Assistant/WritingAssistantFloatingMenu.tsx (hunk 1 above)
Comment (weilirs) [reply src/components/Writing-Assistant/WritingAssistantFloatingMenu.tsx]: fixed

**Comment by weilirs** [comment]
> This is absolutely excellent! Couple of points after running it:
> 
> * We need to handle the case where the generated text flows off the screen. Atm it just goes off the screen unless the window is resized. I think a good solution is to calculate the max size of the component based on the screen height and current position of the component and then once the component reaches this max size just make it have scrollbars in the text it has generated. Thouh overall I think this could be added in a future PR.
> * Another follow on from this will be to think about just like the slash AI command (where no context needs to be selected) which can be used to just write or take in the context of the whole note. Again work for a future PR!

Sure, I'll work on this in future PRs!

**Review by samlhuillier — APPROVED** [review]

**Pull request merged**